feat: int8 support for distance functions

triandco opened the pull request:
Using [AllMiniLM12V2](https://huggingface.co/sentence-transformers/all-MiniLM-L12-v2) to generate embeddings, I found that with a tiny bit of transformation, I can store the embedding as int8 to massively reduce its storage requirement and turn it into f32 during distance calculation (using cosine and dot) with marginal change to the outcome.

I would like to add support for int8 by converting FixedSizeList type of Int8Array to Float32Array during calculation.

The conversion is done with a ``convert_to_floating_point`` on the ``FixedSizeListArrayExt`` trait. In theory, this can also work with other types like uint8, int32, int64. This would make a big difference in storage requirement on the compatible models.

BubbleCal commented:
You probably need to add some tests that verifies we can create vector index on int8 vectors.

It would be nice if you can tell more about your use cases, this looks not efficient for searching.

triandco commented:
@BubbleCal, I'm happy a look at making the index work.

My main use case was to run lance on edge device where storage is limited. Thus I'm looking for way to store as little data as possible, thus the Int8 type. To do this, I did some transformation to project f32 embeddings into int8 space. I ran [BEIR](https://github.com/beir-cellar/beir) on scifact and a couple of other corpus for these transformed vectors and see a minimal change on [ndcg@10](https://en.wikipedia.org/wiki/Discounted_cumulative_gain). Hamming didn't perform very well though.

Again this is an edge device so the quality of search is much more important than speed and efficiency (dataset might be less than 10,000 record). This is why I kinda forgot about building the index. But I think it would be important to make index works fully.

I had a look at the way index is built for non-hamming distance. Look like it was written with the assumption that centroids and residuals are both floating types matching the vector column. In this case, I think I might want to turn transform my index which has an int8 column into a f32 column before the index was built. Would this be an appropriate approach? 

Any thoughts would be much appreciated.

triandco commented:
Added new test for int8 index in ``rust/lance/src/dataset.rs``. 

In order to achieve this we needed to do a couple of thing:
1. Added int8 to f32 conversion for L2, Dot, Cosine during ``ivf::train_ivf_model``
2. Added int8 to f32 conversion for ``kmean::compute_partitions_arrow_array``
3. Allows vector column to be rewritten in f32 during ``residual::compute_residual``

Admittedly these extra conversions mean that it is not as efficient as using floating point for vector column. However, it significantly lower storage requirement. This would increase optionality for down stream consumers allowing them to balance between their storage and speed requirement.

BubbleCal commented:
> Added new test for int8 index in `rust/lance/src/dataset.rs`.
> 
> In order to achieve this we needed to do a couple of thing:
> 
> 1. Added int8 to f32 conversion for L2, Dot, Cosine during `ivf::train_ivf_model`
> 2. Added int8 to f32 conversion for `kmean::compute_partitions_arrow_array`
> 3. Allows vector column to be rewritten in f32 during `residual::compute_residual`
> 
> Admittedly these extra conversions mean that it is not as efficient as using floating point for vector column. However, it significantly lower storage requirement. This would increase optionality for down stream consumers allowing them to balance between their storage and speed requirement.

Thanks for answering!
Yes I think for now the index assumes the data type is floating so it would need more efforts to support integer vectors.
@eddyxu This is fine to me, I think we can erase the casting and implement distance computing with SIMD over integers later, any suggestions?

BubbleCal commented:
Hi @triandco 
it seems there are some lint errors from Clippy, could you fix them then we can continue this?

btw we'd love to hear more about your use cases, looking forward to hearing more in the [Discord](https://discord.com/invite/zMM32dvNtd) community

triandco commented:
Oh thank you @BubbleCal, I didn't know about clippy before. 
I've just ran ``cargo clippy`` on the changed files and fixed the errors. 🙏

BubbleCal commented:
the failed check is not related to this PR, merge it
